Implement torch.linalg.qr

antocuni opened the pull request:
I am opening this PR early to have a place to discuss design issues.
The biggest difference between `torch.qr` and `numpy.linalg.qr` is that the former `torch.qr` takes a boolean parameter `some=True`, while the latter takes a string parameter `mode='reduced'` which can be one of the following:

`reduced`
this is completely equivalent to `some=True`, and both are the default.

`complete`
this is completely equivalent to `some=False`.

`r`
this returns only `r` instead of a tuple `(r, q)`. We have already decided that we don't want different return types depending on the parameters, so I propose to return `(r, empty_tensor)` instead. I **think** that in this mode it will be impossible to implement the backward pass, so we should raise an appropriate error in that case.

`raw`
in this mode, it returns `(h, tau)` instead of `(q, r)`. Internally, `h` and `tau` are obtained by calling lapack's `dgeqrf` and are later used to compute the actual values of `(q, r)`. The numpy docs suggest that these might be useful to call other lapack functions, but at the moment none of them is exposed by numpy and I don't know how often it is used in the real world.
I suppose the implementing the backward pass need attention to: the most straightforward solution is to use `(h, tau)` to compute `(q, r)` and then use the normal logic for `qr_backward`, but there might be faster alternatives.

`full`, `f`
alias for `reduced`, deprecated since numpy 1.8.0

`economic`, `e`
similar to `raw but it returns only `h` instead of `(h, tau). Deprecated since numpy 1.8.0

To summarize: 
  * `reduce`, `complete` and `r` are straightforward to implement. 

  * `raw` needs a bit of extra care, but I don't know how much high priority it is: since it is used rarely, we might want to not support it right now and maybe implement it in the future?

  * I think we should just leave `full` and `economic` out, and possibly add a note to the docs explaining what you need to use instead

/cc @mruberry

antocuni committed:
write two of the tests that we want to pass and a pure-python implementation

mruberry commented:
Thank you for your thorough review, @antocuni.

> `r`
> this returns only `r` instead of a tuple `(r, q)`. We have already decided that we don't want different return types depending on the parameters, so I propose to return `(r, empty_tensor)` instead. I **think** that in this mode it will be impossible to implement the backward pass, so we should raise an appropriate error in that case.

Is this mode actually interesting to support? That is, does not computing `q` save any computation?
 
> `raw`
> in this mode, it returns `(h, tau)` instead of `(q, r)`. Internally, `h` and `tau` are obtained by calling lapack's `dgeqrf` and are later used to compute the actual values of `(q, r)`. The numpy docs suggest that these might be useful to call other lapack functions, but at the moment none of them is exposed by numpy and I don't know how often it is used in the real world.

Sounds like we don't have to implement this, especially not in this first PR. I would only do it if it's convenient.

> `full`, `f`
> alias for `reduced`, deprecated since numpy 1.8.0

Let's not implement this, then. 

> `economic`, `e`
> similar to `raw but it returns only `h`instead of`(h, tau). Deprecated since numpy 1.8.0

Same here; we don't want to implement deprecated NumPy behavior. 

> * `reduce`, `complete` and `r` are straightforward to implement.

Agree with the caveat that implementing `r` only seems interesting if it actually saves computation.

> * `raw` needs a bit of extra care, but I don't know how much high priority it is: since it is used rarely, we might want to not support it right now and maybe implement it in the future?

Agree, per the above. 

> * I think we should just leave `full` and `economic` out, and possibly add a note to the docs explaining what you need to use instead

We probably don't even need the note ;)

antocuni commented:
>> `r`
>> this returns only r instead of a tuple (r, q). We have already decided that we don't want different return types depending on the parameters, so I propose to return (r, empty_tensor) instead. I think that in this mode it will be impossible to implement the backward pass, so we should raise an appropriate error in that case.

> Is this mode actually interesting to support? That is, does not computing q save any computation?

yes, it does. Under the hood both numpy and pytorch computes `r` and `tau` by calling lapack's `GEQRF`. After that, you can compute `q` by calling `GEQRF`, so the `r` mode saves the second call.

Ok, so I'll go and implement only `reduced`, `complete` and `r` in this PR.

antocuni committed:
refactor _qr_helper to take a string argument instead of a bool, will be used by linalg_qr later

antocuni committed:
make linalg_qr the main implementation, and implement qr on top of it

antocuni committed:
introduce a helper function to parse the 'mode' param. Renamed the old arg 'some' into 'reduced' in the internal functions, to match how it is spelled inside 'mode'

antocuni committed:
implement linalg.qr(... mode='r')

antocuni committed:
simplify the 0-elem tensor case. I don't really know why it was written that way, probably a leftover of some refactoring

antocuni committed:
raise an error if we try to backward() after using qr(mode='r')

antocuni committed:
implement the cuda version of _linalg_qr_helper

antocuni committed:
flakeify

antocuni committed:
leftover

antocuni committed:
add a passing test for the out version

antocuni committed:
write docs for torch.linalg.qr

antocuni commented:
removing the "WIP" as I think the branch is ready for review

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -847,22 +852,33 @@ std::tuple<Tensor, Tensor> _qr_helper_cpu(const Tensor& self, bool some) {
 }
 
-std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, bool some) {
+std::tuple<Tensor&,Tensor&> linalg_qr_out(Tensor& Q, Tensor& R, const Tensor& self, std::string mode) {
```
Instead of std::string here and above, can this use const std::string& ?

antocuni replied:
using `const std::string&` doesn't seem supported by `native_functions.yml`.
Moreover, I used it to be consistent with `linalg_norm` which is already using `std::string` for the param `ord`

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -847,22 +852,33 @@ std::tuple<Tensor, Tensor> _qr_helper_cpu(const Tensor& self, bool some) {
   return std::tuple<Tensor&, Tensor&>(Q, R);
 }
 
+std::tuple<Tensor,Tensor> qr(const Tensor& self, bool some) {
```
This is a great remapping of the old function to the new one to avoid code redundancy.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -847,22 +852,33 @@ std::tuple<Tensor, Tensor> _qr_helper_cpu(const Tensor& self, bool some) {
   Tensor Q_tmp, R_tmp;
-  std::tie(Q_tmp, R_tmp) = at::_qr_helper(self, some);
+  std::tie(Q_tmp, R_tmp) = at::_linalg_qr_helper(self, mode);
   Q.resize_as_(Q_tmp).copy_(Q_tmp);
```
Use `resize_output` instead of just resizing so a warning is thrown if the user supplies tensors of the wrong shape

antocuni replied:
b962e414ce

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1447,47 +1449,57 @@ AT_ERROR("qr: MAGMA library not found in "
+    if (compute_q) {
+        // Fix the number of columns of q appropriately
+        q_sizes[self.dim() - 1] = n_columns_q;
+        q_working_copy = at::zeros(q_sizes, self.options());
```
Why is `q_working_copy` filled with zeros here?

antocuni replied:
good catch. Fixed by 490afc403e by using `empty` instead

antocuni replied:
... and re-fixed again in bbe2638289. It seems that this particular codepath was untested on master

mruberry commented on test/test_autograd.py:
```diff
@@ -4904,6 +4904,16 @@ def test_integer_outputs(self):
         self.assertFalse(out.dtype.is_floating_point)
         self.assertFalse(out.requires_grad)
 
+    def test_linalg_qr_r(self):
```
So what's the status of backward with this PR vs PyTorch's existing torch.qr?

mruberry replied:
Add a comment describing why this special autograd function is necessary

antocuni replied:
> So what's the status of backward with this PR vs PyTorch's existing torch.qr?

backwards just works, unless you use `torch.linalg.qr(mode='r')`, in which case it cannot.

> Add a comment describing why this special autograd function is necessary

I'm not sure which function you are talking about. Do you mean to add a comment to explain why the test is necessary?￼

mruberry replied:
Yes, sorry, I was unclear. Please add a comment describing why the test is necessary.

antocuni replied:
ok, fixed by 21375b2855

mruberry commented on test/test_linalg.py:
```diff
@@ -1347,6 +1347,90 @@ def check(equation, operands, regex, exception=RuntimeError):
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_reduced(self, device, dtype):
+        t = torch.randn((7, 5), device=device, dtype=dtype)
```
Test with a batch, too. This test can iterate over an iterable of shapes.

antocuni replied:
I merged most of tests as you suggested in commit bbe2638289.
However, I had to write a separate version to test the batch because numpy doesn't support it: see e2231e9af8

mruberry replied:
Awesome!

mruberry commented on test/test_linalg.py:
```diff
@@ -1347,6 +1347,90 @@ def check(equation, operands, regex, exception=RuntimeError):
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_complete(self, device, dtype):
+        t = torch.randn((7, 5), device=device, dtype=dtype)
+        np_t = t.cpu().numpy()
```
Should this test be combined with the previous one, and the test can enumerate both shapes and the mode?

antocuni replied:
done

mruberry commented on test/test_linalg.py:
```diff
@@ -1347,6 +1347,90 @@ def check(equation, operands, regex, exception=RuntimeError):
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_r(self, device, dtype):
+        t = torch.randn((7, 5), device=device, dtype=dtype)
```
Test with a batch, too

mruberry commented on test/test_linalg.py:
```diff
@@ -1347,6 +1347,90 @@ def check(equation, operands, regex, exception=RuntimeError):
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float)
+    def test_qr_empty(self, device, dtype):
```
Maybe empty shapes can be added to the previous tests when they enumerate shapes?

antocuni replied:
done

mruberry commented on test/test_linalg.py:
```diff
@@ -1347,6 +1347,90 @@ def check(equation, operands, regex, exception=RuntimeError):
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_out(self, device, dtype):
```
Instead of testing out separately, I suggest trying to include it in the tests for each mode.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1944,8 +1945,13 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
+                          std::string mode, const Tensor& q, const Tensor& r){
+  bool compute_q, reduced;
+  std::tie(compute_q, reduced) = at::native::_parse_qr_mode(mode);
+  TORCH_CHECK(compute_q, "linalg_qr_backward: cannot compute backward if mode='r'. "
```
This error message is very good because it makes the problem clear and suggests a remedy to the user. Nice work.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+qr(input, mode='reduced', *, out=None) -> (Tensor, Tensor)
+
+Computes the QR decomposition of a matrix or a batch of matrices :attr:`input`,
+and returns a namedtuple (Q, R) of tensors such that :math:`\text{input} = Q R`
```
I think something is missing between the Q and R in the mathematical expression?

antocuni replied:
[wikipedia](https://en.wikipedia.org/wiki/QR_decomposition) uses the very same notation (albeit without the space):
![image](https://user-images.githubusercontent.com/344507/100471331-5b6ff600-30da-11eb-8e12-606baf89fbe0.png)

mruberry replied:
OK, sounds good.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+Depending on the value of :attr:`mode` this function returns the reduced or
+complete QR factorization. See below for a list of valid modes.
+
+.. warning:: **Differences with** :meth:`~torch.qr`:
```
I think we can remove this section. We should deprecate torch.qr in the future and not worry about describing the differences.

antocuni replied:
as for `svd`, I think that from the user might wonder why there are two functions with the same name, especially if he/she uses the "search" box in the docs. So being able to immediately compare them is useful.

But I am fine to delete it if you really want.

mruberry replied:
You make a good point and it'd be nice if our docs could put "deprecated" next to the name of the functions when they're searched for, so the search would show something like:

```
torch.linalg.qr
torch.qr (deprecated)
```

Adding the deprecation note to torch.qr will be half the battle. cc @heitorschueroff who's also going to help us review torch.linalg consistency before the 1.8 release. 

Let's delete this for now and consider what we want to do systematically before the 1.8 release. That is, if we do want warnings or notes like this we should put them on all the new linear algebra functions that replace existing functions, too.

antocuni replied:
to be consistent with what we decided to do in #45562 , I move the "differences" box to `torch.qr`, see commit 723c4d5ca2.
However, I noticed that there is a problem with how sphinxs renders ``:meth:`~torch.linalg.qr` ``: see the following pic:
![image](https://user-images.githubusercontent.com/344507/101518536-53d60a00-3982-11eb-94f8-5d5508cd6aa1.png)

as you can see, the reference to `torch.linalg.qr` is rendered as just `qr`. So, as a workaround I prefixed it with a fixed-font ``torch.linalg.``, but it's ugly.

I think that this misbehavior should be fixed in in another PR, though.

rgommers replied:
`:meth:` seems wrong. I would expect it to work right if you do `` `torch.linalg.qr` `` (just single backticks). And otherwise it should be `:func:` right?

Cc @mattip as our resident expert.

antocuni replied:
The single backticks don't work (they are just rendered in *italics*). I also tried to use `:func:` instead of `:meth:` but I get the same rendering as the picture above

mattip replied:
These kind of deliberations can be resolved by looking at the objects stored in https://github.com/pytorch/pytorch.github.io/blob/site/docs/master/objects.inv (or https://pytorch.org/docs/master/objects.inv) with [sphobjinv](https://github.com/bskinn/sphobjinv) like this:
```
$ pip install sphobjinv
$ wget https://pytorch.org/docs/master/objects.inv
$ sphobjinv suggest objects.inv linalg.qr
:py:method:`torch.Tensor.qr`
:py:module:`torch.linalg`
:py:function:`torch.linalg.cholesky`
:py:function:`torch.linalg.cond`
:py:function:`torch.linalg.det`
:py:function:`torch.linalg.eigh`
:py:function:`torch.linalg.eigvalsh`
:py:function:`torch.linalg.matrix_rank`
:py:function:`torch.linalg.norm`
:py:function:`torch.linalg.tensorinv`
:py:function:`torch.linalg.tensorsolve`
:py:function:`torch.qr`
:std:doc:`generated/torch.qr`
:std:label:`generated/torch.qr:torch.qr`
:std:doc:`linalg`
:std:label:`linalg:functions`
:std:label:`linalg:torch.linalg`
```

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+
+               - ``some=False`` is equivalent of ``mode='complete'``.
+
+             **Differences with** ``numpy.linalg.qr``:
```
For the differences with NumPy, instead of a warning we can probably have a note like:

"Unlike NumPy's linalg.qr, this function always returns a tuple of two tensors. When \`\`mode='r'\`\`, the \`Q\` tensor is an empty tensor."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+               tensor.
+
+
+.. warning::
```
This can be a note instead of a warning, too.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+          If you plan to backpropagate through QR, note that the current backward implementation
+          is only well-defined when the first :math:`\min(input.size(-1), input.size(-2))`
+          columns of :attr:`input` are linearly independent.
+          This behavior will propably change once QR supports pivoting.
```
Let's just say, "This behavior may change in the future."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+          columns of :attr:`input` are linearly independent.
+          This behavior will propably change once QR supports pivoting.
+
+.. note:: precision may be lost if the magnitudes of the elements of :attr:`input`
```
I wonder if this note is precise enough to be helpful? Do you think it's worth keeping?

antocuni replied:
honestly, I don't know. This was just a copy&paste from the docs of `torch.qr`

mruberry replied:
Let's remove it for now.

antocuni replied:
removed by both `qr` and `linalg.qr` in commit c18410e329

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -184,3 +184,94 @@
+.. note:: precision may be lost if the magnitudes of the elements of :attr:`input`
+          are large
+
+.. note:: While it should always give you a valid decomposition, it may not
```
"This function may produce different decompositions on different device types and different platforms."

This sentence is both simpler and more precise, since CUDA, for example, doesn't use LAPACK.

albanD replied:
I would be in favor of keeping LAPACK/MAGMA mention here as it makes it clear to the user that this is not our implementation choice but a quirk coming from third party libraries we are using.
And they should expect such values to change when they change these libraries even with the same version of pytorch.

mruberry replied:
If we want to mention the mathematical library used for operators, which is a fine idea, then I would just add a sentence after the proposed sentence like: "This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs." Also, aren't there multiple versions of LAPACK?

albanD replied:
> Also, aren't there multiple versions of LAPACK?

Ho yes there are. But I don't think we have a way to print to the user which blas/lapack lib they're using.

antocuni replied:
I tried to summarize it this way:
```
.. note:: This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs,
          and may produce different decompositions on different device types
          and different platforms, depending on the precise version of the
          underlying library.
```

rgommers replied:
Maybe add "(valid)" before "decompositions"? This note makes it seem like a flaw in the implementation, but the actual message here to the user is that numerical values in`Q` and `R` may be wildly different, and that relying on those values is user error - only the mathematical properties of the returned matrices should be relied on.

antocuni replied:
good idea. I implemented ralph's suggestion in commit aba9124d8e, which also updates ``torch.qr`` to use the very same sentence

mruberry reviewed (commented):
Hey @antocuni!

This is looking pretty good. I made some suggestions and asked a few questions. Looking forward to your responses!

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1944,8 +1945,13 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
+  bool compute_q, reduced;
+  std::tie(compute_q, reduced) = at::native::_parse_qr_mode(mode);
+  TORCH_CHECK(compute_q, "linalg_qr_backward: cannot compute backward if mode='r'. "
+                         "Please use torch.linalg.qr(..., mode='reduced')");
```
nit: wouldn't complete work here are well? Or reduced is the one to use to get the same r result?

antocuni replied:
you must use `reduced` to get the same result for the `r` matrix. If you use `complete`, backwards works but you get a different result

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6276,20 +6276,13 @@
-  dispatch:
-    CPU: _qr_helper_cpu
-    CUDA: _qr_helper_cuda
+    Math: qr
```
AFAIK This function has a custom dispatch and so is not a composite function. This means that it should have a gradient formula defined in derivatives.yaml otherwise it will be marked as non-differentiable automatically.

Could you check the generated code for ` qr(` in `torch/csrc/autograd/generated/VariableTypeEverything.cpp` please to make sure it is still properly differentiable?

antocuni replied:
> Could you check the generated code for qr( in torch/csrc/autograd/generated/VariableTypeEverything.cpp please to make sure it is still properly differentiable?

I'm not very familiar with that code so I'm not sure what exactly I should check. But there are tests which check that `qr` is differentiable and they all pass.

The trick is that now `torch.qr` and `torch.lingalg.qr` are thin wrappers around `_qr_helper`: so `derivatives.yml` contains just the formula for the latter.

mruberry replied:
Yep; I think autograd is working as expected.

antocuni replied:
ok, marking this thread as resolved then

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6276,20 +6276,13 @@
 
 - func: qr.Q(Tensor self, bool some=True, *, Tensor(a!) Q, Tensor(b!) R) -> (Tensor(a!) Q, Tensor(b!) R)
```
@mruberry if this is an out variant, should it be renamed `qr.out`? The codegen has quite a few special case expecting the out variants to be named this way.

antocuni replied:
I think that the convention is to use the name of the first different argument after the dot. So most functions are called `.out` because the argument is called `out`. But here we have two outputs so they are named differently.

I am almost certain to have read this somewhere in the docs but I can no longer find it now.

mruberry replied:
This is consistent with the current qr:

https://github.com/pytorch/pytorch/blob/7df84452423f44ebe1db40a2e3463066bf954f95/aten/src/ATen/native/native_functions.yaml#L6343

But it should be fine to change it to `qr.out`, too, so we might as well.

antocuni replied:
ok, I renamed the linalg version to `linalg_qr.out`, but left the old one unchanged. See 2a29fe53c5

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8966,6 +8959,26 @@
+  dispatch:
+    DefaultBackend: linalg_qr
+
+- func: linalg_qr.Q(Tensor self, str mode='reduced', *, Tensor(a!) Q, Tensor(b!) R) -> (Tensor(a!) Q, Tensor(b!) R)
```
Same here, I would expect this one to be `linalg_qr.out` as it is an out variant.

antocuni committed:
use resize_output to get a warning if we pass an out tensor of the wrong size, and fix the test

antocuni committed:
no need to use zeros, empty is enough

antocuni committed:
refactor the tests and merge them into two big ones. This actually catches a bug because mode='complete' with empty arrays was untested. This partially reverts commit 8653e81605 and explains why I could use at::empty instead of at::eye without breaking anything: this particular case was untested even on master!

antocuni committed:
test the batched version

antocuni committed:
tweak docs

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-qr

antocuni committed:
fix test_namedtuple_return_api.py

antocuni committed:
add a comment explaining why we need such a test

antocuni committed:
move the box with differences between torch.linalg.qr and torch.qr into the latter

antocuni committed:
remove this note

antocuni committed:
improve this note, and use the very same wording also in torch.qr

antocuni committed:
rename this to linalg_qr.out, it looks cleaner

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-qr

antocuni commented:
@mruberry I think I addressed most of the reviews. The only remaining bit is the small quirk in the docs as shown by my [earlier comment](https://github.com/pytorch/pytorch/pull/47764#discussion_r538629779)

antocuni committed:
test_qr was ported from test_torch.py to test_linalg.py in master. Move it closer to the qr tests introduced by this branch, and rename the existing test_qr into test_qr_vs_numpy

antocuni committed:
these tests were killed by upstream/master, not sure why they survived the merge

mruberry commented:
> @mruberry I think I addressed most of the reviews. The only remaining bit is the small quirk in the docs as shown by my [earlier comment](https://github.com/pytorch/pytorch/pull/47764#discussion_r538629779)

Cool! Looks like the team is working that quirk out, too. It is odd. We don't need to block this PR on a resolution, however. It's OK to just put torch.linalg.qr in as a string if we have to.

The build errors on this PR still seem real, however. For example:

```
Dec 08 21:28:16 /var/lib/jenkins/workspace/aten/src/ATen/native/hip/BatchLinearAlgebra.hip:1785:3: error: unknown type name 'magma_int_t'
```

antocuni committed:
we renamed _qr_helper to _linalg_qr_helper, but it's fine since it's an operator which is supposed to be called internally

antocuni committed:
blindly try fix compilation on ROCm, by moving the magma_int_cast back into an #ifdef USE_MAGMA block

antocuni committed:
fix test_overrides

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-qr

antocuni commented:
> The build errors on this PR still seem real, however. For example:
[cut]

@mruberry sorry, I didn't see those failures. Everything seems to be fixed now!

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-qr

mruberry commented:
Test failures are still real. Does this need to be added to the list of ops that work with complex autograd? 

See here:

https://github.com/pytorch/pytorch/blob/6c1b405a3bc5392081569f1530d74e8459e0e211/tools/autograd/gen_variable_type.py#L68

antocuni committed:
git merge upstream/master

antocuni committed:
the main differentiable function is linalg_qr nowadays

antocuni commented:
> Test failures are still real. Does this need to be added to the list of ops that work with complex autograd?

yes, the problem was that in this branch the main function is `linalg_qr`, and `qr` is implemented on top of it. I pushed the fix, waiting for the CI jobs

antocuni committed:
remove git merge leftover

antocuni commented:
@mruberry it seems that all CI tests have passed, so the PR is ready for a new review round

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1005,22 +1009,34 @@ std::tuple<Tensor, Tensor> _qr_helper_cpu(const Tensor& self, bool some) {
-std::tuple<Tensor,Tensor> qr(const Tensor& self, bool some) {
+std::tuple<Tensor,Tensor> linalg_qr(const Tensor& self, std::string mode) {
   TORCH_CHECK(self.dim() >= 2,
               "self should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
```
"self" -> "qr input"

The name "self" is almost exclusively internal, and the error message should reflect the documentation, which says "input". The error message should also include the name of the function so people can understand where it's coming from.

mruberry replied:
(I understand this is an existing problem, but while we're fixing qr we might as well improve it.)

antocuni replied:
fixed by commit 8201ec9123 in PR #50046

mruberry commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -192,16 +192,35 @@ static inline Tensor _move_to_end(const Tensor& self, IntArrayRef axes) {
+    compute_q = false;
+    reduced = true; // this is actually irrelevant in this mode
+  } else {
+    TORCH_CHECK(false, "Unrecognized mode '", mode, "'");
```
"qr received unrecognized mode MODE but expected one of reduced (default), r, or complete"

antocuni replied:
8201ec9123

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10169,6 +10162,27 @@
+  python_module: linalg
+  variants: function
+  dispatch:
+    DefaultBackend: linalg_qr_out
```
What's the distinction between DefaultBackend and math dispatch here?

In practice we could probably make all these dispatches CPU, CUDA:, since ultimately these functions are just wrappers around _linalg_qr_helper. I'm not sure there's much value in making them Math or DefaultBackend.

antocuni replied:
I'm not sure to understand the question. In this branch `linalg_qr` and `linalg_qr_out` are the "main" functions: they have a corresponding entry in `derivatives.yaml` and thus they require `DefaultBackend`, if I understand correctly how the `dispatch:` mechanism works.
On the other hand `qr` and `qr_out` are implemented in terms of `linalg_qr` and thus they use `dispatch: Math`.

For the record, I tried to use `Math` for `linalg_qr_out` but as expected I get this compilation error:
```
AssertionError: There's a formula for linalg_qr_out(or its functional variant) in derivatives.yaml. It's required to add a dispatch section for it with explicit supported backends e.g CPU/CUDA or DefaultBackend in native_functions.yaml. Please see https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#choosing-the-right-dispatch-keyword for instructions to choose the right dispatch keyword.
```

mruberry replied:
Interesting, I didn't realize that we were tying "out" variants to derivatives.yaml.

@albanD @smessmer @ailzhang  - we should probably update the developer FAQ (https://github.com/pytorch/pytorch/wiki/Developer-FAQ) with a description of when to use each dispatch and how entries are associated in autograd.yaml with a couple examples.

albanD replied:
I don't think we should have a formula for any ".out" variant. And the codegen would prevent that I think if you use the naming convention with the `.`

mruberry commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -32,6 +32,158 @@
     # Internal, profiler-specific ops
     ("profiler::_call_end_callbacks_on_jit_fut*", datetime.date(9999, 1, 1)),
     ("profiler::_record_function_enter", datetime.date(9999, 1, 1)),
+    ("tensorexpr::Group", datetime.date(2020, 9, 9)),
```
This seems like a lot of unrelated changes? Was there a merge issue?

antocuni replied:
yes, I think I did a bad merge. They were killed by 49c9994fb7 and probably I reintroduced them by mistake.
Killed them again in commit 609b3fdb8d

mruberry commented on test/test_autograd.py:
```diff
@@ -4980,6 +4980,17 @@ def assert_only_first_requires_grad(res):
                                          return_counts=return_counts)
                     assert_only_first_requires_grad(res)
 
+    def test_linalg_qr_r(self):
```
This is a great test. Let's move it to test_linalg.py where the other qr tests live.

In the future I expect test_autograd.py to test autograd system-specific things.

antocuni replied:
261133cdfc

mruberry commented on test/test_autograd.py:
```diff
@@ -4980,6 +4980,17 @@ def assert_only_first_requires_grad(res):
+        # linalg_qr_backward complains cleanly in that case.
+        inp = torch.randn((5, 7), requires_grad=True)
+        q, r = torch.linalg.qr(inp, mode='r')
+        assert q.shape == (0,)  # empty tensor
```
Instead of a naked assert maybe assertEqual here?

antocuni replied:
fixed by commit 9299f87186, together with a bunch of other naked asserts

mruberry commented on test/test_linalg.py:
```diff
@@ -2716,6 +2716,135 @@ def test_old_matrix_rank(self, device, dtype):
+            self.assertEqual(A_, np.matmul(Q_, R_))
+
+            # Check2: A = QR (with out)
+            Q_out, R_out = torch.full_like(Q, math.nan), torch.full_like(R, math.nan)
```
Filling the out tensors with NaNs here is great

mruberry commented on test/test_linalg.py:
```diff
@@ -2716,6 +2716,135 @@ def test_old_matrix_rank(self, device, dtype):
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_vs_numpy(self, device, dtype):
```
This test is great

mruberry commented on test/test_linalg.py:
```diff
@@ -2716,6 +2716,135 @@ def test_old_matrix_rank(self, device, dtype):
+                all_r.append(r)
+            return np.array(all_q), np.array(all_r)
+
+        t = torch.randn((3, 7, 5), device=device, dtype=dtype)
```
This is cool; let's extend this with a test for zero batch size and two batch dimensions.

The zero batch size test can be separate from this test since it wouldn't involve comparing with NumPy.

Reshaping or flattening/unflattening should make the multiple batch dimensions comparison possible vs. NumPy.

antocuni replied:
> This is cool; let's extend this with a test for zero batch size and two batch dimensions.

do you mean a shape of `(0, 7, 5)`, or something else? I'm not sure what it's supposed to do in that case, though.

> Reshaping or flattening/unflattening should make the multiple batch dimensions comparison possible vs. NumPy.

are you suggesting to rewrite `np_qr_batched` using reshaping/flattening instead of the explicit loop which I wrote?

mruberry replied:
I was thinking of a shape like (0, 7, 5), yep. Is the issue that we don't support a batch dim of size zero? 

> are you suggesting to rewrite np_qr_batched using reshaping/flattening instead of the explicit loop which I wrote?

"Rewrite" sounds like a lot. Just a test using multiple batch dimensions that compares against NumPy would be great.

mruberry commented on test/test_linalg.py:
```diff
@@ -2716,6 +2716,135 @@ def test_old_matrix_rank(self, device, dtype):
+
+        t = torch.randn((3, 7, 5), device=device, dtype=dtype)
+        np_t = t.cpu().numpy()
+        for mode in ['reduced', 'complete']:
```
What about mode "r" here?

antocuni replied:
test added by commit 8d5c38511d

mruberry commented on test/test_linalg.py:
```diff
@@ -2716,6 +2716,135 @@ def test_old_matrix_rank(self, device, dtype):
+                assert r2 is out[1]
+                self.assertEqual(q2, q)
+                self.assertEqual(r2, r)
+
```
Is there a test for error cases like receiving a 1D tensor as input or a mode that's not supported?

antocuni replied:
good idea, I added a test in cdac407ae6

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6672,18 +6672,30 @@ def merge_dicts(*dicts):
 If :attr:`some` is ``True``, then this function returns the thin (reduced) QR factorization.
 Otherwise, if :attr:`some` is ``False``, this function returns the complete QR factorization.
 
+.. warning:: ``torch.qr`` is deprecated. Please use ``torch.linalg.`` :meth:`~torch.linalg.qr`
```
:func:, not :meth:, since torch.linalg.qr is a function

mruberry replied:
This is has some funky formatting:

<img width="720" alt="Screen Shot 2020-12-24 at 9 01 38 PM" src="https://user-images.githubusercontent.com/38511765/103120127-42039080-462b-11eb-8fd4-af482da50af6.png">

Can we make the entire name a reference?

antocuni replied:
partially fixed by a5570d7797,

> This is has some funky formatting:

I know, I have already mentioned it in a previous comment: https://github.com/pytorch/pytorch/pull/47764#discussion_r538629779

The problem is that `` :meth:`~torch.linalg.qr` `` is rendered as simply `qr` by sphinx: what I did is a workaround but I agree it should be properly solved eventually.

mruberry replied:
I agree this is a reasonable workaround.

bskinn replied:
Late here, this may already be solved, but -- does using ```:func:`torch.linalg.qr` ``` (i.e., removing the `~`) not render the link the way you want?

The `~` [tells Sphinx](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-syntax) to only render the simple object name, rather than the fully-qualified name that (I think) you want here.

mruberry replied:
Never too late. Thanks for fixing this, @bskinn!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6672,18 +6672,30 @@ def merge_dicts(*dicts):
 Otherwise, if :attr:`some` is ``False``, this function returns the complete QR factorization.
 
+.. warning:: ``torch.qr`` is deprecated. Please use ``torch.linalg.`` :meth:`~torch.linalg.qr`
+             instead, which provides a better compatibility with
```
I think we can remove everything after "instead," and let users discover what linalg.qr does on its page.

antocuni replied:
785136fb37

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6672,18 +6672,30 @@ def merge_dicts(*dicts):
+
+             * ``torch.linalg.qr`` takes a string parameter ``mode`` instead of ``some``:
+
+               - ``some=True`` is equivalent of ``mode='reduced'``: both are the
```
This is a cool update guide.

cc @heitorschueroff

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6672,18 +6672,30 @@ def merge_dicts(*dicts):
-.. note:: While it should always give you a valid decomposition, it may not
-          give you the same one across platforms - it will depend on your
-          LAPACK implementation.
+.. note:: This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs,
```
This is a much better note than the old one

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6672,18 +6672,30 @@ def merge_dicts(*dicts):
-          LAPACK implementation.
+.. note:: This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs,
+          and may produce different (valid) decompositions on different device types
+          and different platforms, depending on the precise version of the
```
"on different device types and different platforms," -> "on different device types or different platforms."

And then we can lose the rest of the sentence after the comma since it's redundant with "different platforms."

antocuni replied:
785136fb37

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2141,7 +2147,7 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
   TORCH_CHECK(
-      ((m <= n && (!some)) || some),
+      ((m <= n && (!reduced)) || reduced),
       "The derivative is not implemented when nrows > ncols and complete QR. ");
```
Let's improve this error message while we're here. The phrase "and complete QR" is odd and vague. Can we relate it more directly to the modes?

antocuni replied:
done by dc24806a8a. I also added a test and made the two error messages more consistent

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -620,3 +620,82 @@
+          Backpropagation is not supported for ``mode='r'``. Use ``mode='reduced'`` instead.
+
+          If you plan to backpropagate through QR, note that the current backward implementation
+          is only well-defined when the first :math:`\min(input.size(-1), input.size(-2))`
```
"is only well-defined" is a little vague. Maybe we should say something like:

"Backpropagation is also not supported if the first :math:\`\min(input.size(-1), input.size(-2))\` columns of any matrix in :attr:\`input\` are not linearly independent. While no error will be thrown when this occurs the values of the "gradient" produced may be anything."

antocuni replied:
87efe7df51

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -620,3 +620,82 @@
+
+.. note:: This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs,
+          and may produce different (valid) decompositions on different device types
+          and different platforms, depending on the precise version of the
```
As above, we can probably kill this note after "platforms.".

antocuni replied:
87efe7df51

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -620,3 +620,82 @@
+
+Keyword args:
+    out (tuple, optional): tuple of `Q` and `R` tensors
+                satisfying :code:`input = torch.matmul(Q, R)`.
```
The tensors provided here don't have to satisfy this relationship. I think line 670 can be removed.

antocuni replied:
87efe7df51

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -620,3 +620,82 @@
+                satisfying :code:`input = torch.matmul(Q, R)`.
+                The dimensions of `Q` and `R` are :math:`(*, m, k)` and :math:`(*, k, n)`
+                respectively, where :math:`k = \min(m, n)` if :attr:`mode` is `'reduced'` and
+                :math:`k = m` if :attr:`mode` is `'complete'`.
```
What if the mode is 'r'?

antocuni replied:
fixed by 16d82412cf. I refactored&simplified the statement, since the very same information is given just a couple of lines above.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -620,3 +620,82 @@
+
+Example::
+
+    >>> a = torch.tensor([[12., -51, 4], [6, 167, -68], [-4, 24, -41]])
```
Showing mode r and how it returns an empty tensor would be interesting, too

antocuni replied:
d7e0fca3dc

mruberry reviewed (commented):
Hey @antocuni!

This is looking pretty good. I've made some more inline comment, mostly to extend some of the tests and documentation.

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-qr

antocuni committed:
these lines were killed by commit 49c9994fb7 but were somehow reintroduced due to a merge issue. Kill them again

antocuni commented:
> Hey @antocuni!
> 
> This is looking pretty good. I've made some more inline comment, mostly to extend some of the tests and documentation.

Thank you @mruberry! I have fixed the problem with `check_backward_compatibility` which was probably due to a bad merge, and merged the latest `upstream/master`. Hopefully the CI will still be green after that.

If you agree, I propose to merge it now, and I'll be happy to fix all the remaining details in a follow-up PR.

mruberry commented:
> If you agree, I propose to merge it now, and I'll be happy to fix all the remaining details in a follow-up PR.

Sounds good. That will save us from merge headaches going forward. We should make these remaining tweaks soon, though.

antocuni commented:
> Sounds good. That will save us from merge headaches going forward. We should make these remaining tweaks soon, though.

sure, I'll do it next week as soon as I am back at work
